src/box/relay.cc

```diff
@@ -1266,27 +1344,43 @@ relay_send_row(struct xstream *stream, struct xrow_header *packet)
 			say_warn("injected broken lsn: %lld",
 				 (long long) packet->lsn);
 		}
-		if (iproto_type_is_promote_request(packet->type)) {
-			struct synchro_request req;
-			xrow_decode_synchro(packet, &req);
-			/*
-			 * PROMOTE/DEMOTE should be sent only after
-			 * corresponding RAFT term was already sent.
-			 * We assume that PROMOTE/DEMOTE will arive after RAFT
-			 * term, otherwise something might break.
-			 */
-			while (relay->sent_raft_term < req.term) {
-				if (fiber_is_cancelled()) {
-					diag_set(FiberIsCancelled);
-					diag_raise();
-				}
-				cbus_process(&relay->tx_endpoint);
-				if (relay->sent_raft_term >= req.term)
-					break;
-				fiber_yield();
-			}
-		}
 		relay_send(relay, packet);
-		relay->is_sending_tx = !packet->is_commit;
 	}
+
+	rlist_create(tx);
+	lsregion_gc(&relay->lsregion, relay->lsr_id);
+}
+
+static void
+relay_process_row(struct xstream *stream, struct xrow_header *packet)
+{
+	struct relay *relay = container_of(stream, struct relay, stream);
+	static __thread int64_t read_tsn = 0;
+	static __thread struct rlist current_tx;
```

Comment:
I would suggest to avoid global variables where possible, and move them into `struct relay`.